stat/mds/mds.go

```diff
@@ -0,0 +1,88 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mds
+
+import (
+	"math"
+	"sort"
+
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/mat"
+)
+
+// TorgersonScaling converts a dissimilarity matrix to a matrix containing
+// Euclidean coordinates. TorgersonScaling places the coordinates in dst and
+// returns it.
+//
+// If dst is nil, a new mat.Dense is allocated. If dst is not a zero matrix,
+// the dimensions of dst and dis must match otherwise TorgersonScaling will panic.
+// The dis matrix must be square or TorgersonScaling will panic.
+func TorgersonScaling(dst *mat.Dense, dis mat.Symmetric) *mat.Dense {
+	// https://doi.org/10.1007/0-387-28981-X_12
+
+	n := dis.Symmetric()
+	if dst == nil {
+		dst = mat.NewDense(n, n, nil)
+	} else if r, c := dst.Dims(); !dst.IsZero() && (r != n || c != n) {
+		panic(mat.ErrShape)
+	}
+
+	b := mat.NewSymDense(n, nil)
+	for i := 0; i < n; i++ {
+		for j := i; j < n; j++ {
+			v := dis.At(i, j)
+			v *= v
+			b.SetSym(i, j, v)
+		}
+	}
+	c := mat.NewSymDense(n, nil)
+	for i := 0; i < n; i++ {
+		s := -1 / float64(n)
+		c.SetSym(i, i, 1+s)
+		for j := i + 1; j < n; j++ {
+			c.SetSym(i, j, s)
+		}
+	}
+	dst.Product(c, b, c)
+	for i := 0; i < n; i++ {
+		for j := i; j < n; j++ {
+			b.SetSym(i, j, -0.5*dst.At(i, j))
+		}
+	}
+
+	var ed mat.EigenSym
+	ed.Factorize(b, true)
+	dst.EigenvectorsSym(&ed)
+	vals := ed.Values(nil)
+	sort.Sort(byValues{
```

Comment:
@vladimir-ch This sort is necessary to get the Eigen values to be ordered appropriately. Is there a better way of doing this? Other ED implementations return the values and vectors ordered such that the values are decreasing, but our documentation is silent on the ordering.

Reply:
Yes, I think this would be the way to do it but EigenSym uses Dsyev which promises to return the eigenvalues in ascending order, so this sorting is not necessary. The documentation should be updated.

Reply:
Great. So I can just do a order reverse. I'll make that change soon.

Also, #461.

Reply:
Done